### PR TITLE
Add PutObject checksums

### PR DESCRIPTION
**Pull request by klauspost**
Single part, checksums must be done manually for now until we have trailing headers.

For multipart we add CRC32C if we don't already send a checksum and we read the content before sending.

Also serves as test for https://github.com/minio/minio/pull/15433 - but should be backwards compatible.

# AWS S3 tests

## Single Part
```
{"args":{"bucketName":"minio-go-test-gyprb0a1v4l9x05q","objectName":"i3d5m60tguye3hkgxe1o3gcijmynm0","opts":"minio.PutObjectOptions{UserMetadata: metadata, Progress: progress}"},"duration":246249,"function":"PutObject(bucketName, objectName, reader,size, opts)","name":"minio-go: testPutObjectWithChecksums","status":"PASS"}
```

# Multipart

```
These send checksum:
	testPutObjectWithMetadata()
	testPutObjectReadAt()
	testPutObjectsUnknownV2()
	testPutObjectNoLengthV2()

go build functional_tests.go   && ./functional_tests
CHECKSUM: 1yLc4A==
{"args":{"bucketName":"minio-go-test-duyyipdmn2jt2axy","metadata":{"Content-Type":["custom/contenttype"],"X-Amz-Meta-CustomKey":["extra  spaces  in   value"]},"objectName":"zn2pdxznyk6rg46yx2jfifgtemt2u9","opts":"minio.PutObjectOptions{UserMetadata: metadata, Progress: progress}"},"duration":71907,"function":"PutObject(bucketName, objectName, reader,size, opts)","name":"minio-go: testPutObjectWithMetadata","status":"PASS"}
CHECKSUM: 1yLc4A==
{"args":{"bucketName":"minio-go-test-d9smpwyxtqnxcxug","objectContentType":"binary/octet-stream","objectName":"fjuqzickiy94a5i2e1od5exi41bvyh","opts":"objectContentType"},"duration":87869,"function":"PutObject(bucketName, objectName, reader, opts)","name":"minio-go: testPutObjectReadAt","status":"PASS"}
CHECKSUM: 73DecA==
CHECKSUM: 73DecA==
CHECKSUM: 73DecA==
CHECKSUM: 73DecA==
{"args":{"bucketName":"minio-go-test-zlq0e6ayw0hgzmhh","objectName":"minio-go-test-zlq0e6ayw0hgzmhhunique4","opts":"","size":""},"duration":3278,"function":"PutObject(bucketName, objectName, reader,size,opts)","name":"minio-go: testPutObjectsUnknownV2","status":"PASS"}
CHECKSUM: LEgalg==
{"args":{"bucketName":"minio-go-test-f5fssd4wb3opahi0","objectName":"minio-go-test-f5fssd4wb3opahi0unique","opts":"","size":135266304},"duration":46122,"function":"PutObject(bucketName, objectName, reader, size, opts)","name":"minio-go: testPutObjectNoLengthV2","status":"PASS"}	
```

Seems we are compatible with S3.

### CODE REVIEW

**Review by harshavardhana — APPROVED** [review]

File: api-put-object.go [hunk 1]
```diff
@@ -299,11 +300,17 @@ func (c *Client) putObjectMultipartStreamNoLength(ctx context.Context, bucketNam
 	if err != nil {
 		return UploadInfo{}, err
 	}
+
+	if !opts.SendContentMd5 {
+		opts.UserMetadata = map[string]string{"X-Amz-Checksum-Algorithm": "CRC32C"}
```
Comment (vadmeste) [inline api-put-object.go:305]: I think you should not initialize a new UserMetadata here if opts.userMetadata is not nil, but just only add if it is already initialized.

This is called just before newUploadID() and usually user metadata are initialized in this stage, NewMultipartUpload().

File: api-put-object.go [hunk 2]
```diff
@@ -377,15 +392,26 @@ func (c *Client) putObjectMultipartStreamNoLength(ctx context.Context, bucketNam
...
 	// Sort all completed parts.
 	sort.Sort(completedParts(complMultipartUpload.Parts))
 
-	uploadInfo, err := c.completeMultipartUpload(ctx, bucketName, objectName, uploadID, complMultipartUpload, PutObjectOptions{})
+	opts = PutObjectOptions{}
+	if len(crcBytes) > 0 {
+		// Add hash of hashes.
+		crc.Reset()
+		crc.Write(crcBytes)
+		opts.UserMetadata = map[string]string{"X-Amz-Checksum-Crc32c": base64.StdEncoding.EncodeToString(crc.Sum(nil))}
```
Comment (vadmeste) [inline api-put-object.go:412]: Let's just not override if opts.UserMetadata is not nil, just add if that's the case

**Review by vadmeste — COMMENTED** [review]
Few comments

File: api-put-object-multipart.go [hunk 3]
```diff
@@ -171,15 +185,25 @@ func (c *Client) putObjectMultipartNoStream(ctx context.Context, bucketName, obj
...
 	// Sort all completed parts.
 	sort.Sort(completedParts(complMultipartUpload.Parts))
-
-	uploadInfo, err := c.completeMultipartUpload(ctx, bucketName, objectName, uploadID, complMultipartUpload, PutObjectOptions{})
+	opts = PutObjectOptions{}
+	if len(crcBytes) > 0 {
+		// Add hash of hashes.
+		crc.Reset()
+		crc.Write(crcBytes)
+		opts.UserMetadata = map[string]string{"X-Amz-Checksum-Crc32c": base64.StdEncoding.EncodeToString(crc.Sum(nil))}
```
Comment (klauspost) [inline api-put-object-multipart.go:204]: Same here I guess...

**Review by harshavardhana — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]
LGTM

**Comment by klauspost** [comment]
Ping @harshavardhana

**Pull request merged**